#!/usr/bin/env python3
# server/calculator_server.py

import grpc
from concurrent import futures
import time
import sys
import os
from datetime import datetime

# æ·»åŠ é¡¹ç›®æ ¹ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from generated import calculator_pb2
from generated import calculator_pb2_grpc


class CalculatorServicer(calculator_pb2_grpc.CalculatorServicer):
    """CalculatoræœåŠ¡å®ç°"""

    def Add(self, request, context):
        """ä¸€å…ƒRPCï¼šç®€å•åŠ æ³•"""
        print(f"[{datetime.now()}] æ”¶åˆ°åŠ æ³•è¯·æ±‚: {request.num1} + {request.num2}")

        result = request.num1 + request.num2

        return calculator_pb2.AddResponse(
            result=result,
            message=f"è®¡ç®—ç»“æœ: {request.num1} + {request.num2} = {result}",
            timestamp=int(time.time() * 1000),
            request_id=request.request_id or "unknown"
        )

    def AddStream(self, request, context):
        """æœåŠ¡ç«¯æµå¼RPCï¼šè¿”å›è®¡ç®—æ­¥éª¤"""
        print(f"[{datetime.now()}] æ”¶åˆ°æµå¼åŠ æ³•è¯·æ±‚: {request.num1} + {request.num2}")

        # æ­¥éª¤1ï¼šæ¥æ”¶æ•°å­—
        yield calculator_pb2.CalculationStep(
            step="1/4",
            current_result=0,
            message=f"æ”¶åˆ°æ•°å­—: {request.num1} å’Œ {request.num2}",
            timestamp=int(time.time() * 1000)
        )

        time.sleep(0.5)  # æ¨¡æ‹Ÿå¤„ç†æ—¶é—´

        # æ­¥éª¤2ï¼šå¼€å§‹è®¡ç®—
        yield calculator_pb2.CalculationStep(
            step="2/4",
            current_result=request.num1,
            message=f"å¼€å§‹è®¡ç®—: ç¬¬ä¸€ä¸ªæ•°å­—æ˜¯ {request.num1}",
            timestamp=int(time.time() * 1000)
        )

        time.sleep(0.5)

        # æ­¥éª¤3ï¼šç»§ç»­è®¡ç®—
        current_sum = request.num1
        yield calculator_pb2.CalculationStep(
            step="3/4",
            current_result=current_sum,
            message=f"åŠ ä¸Šç¬¬äºŒä¸ªæ•°å­—: {request.num2}",
            timestamp=int(time.time() * 1000)
        )

        time.sleep(0.5)

        # æ­¥éª¤4ï¼šå®Œæˆè®¡ç®—
        result = request.num1 + request.num2
        yield calculator_pb2.CalculationStep(
            step="4/4",
            current_result=result,
            message=f"è®¡ç®—å®Œæˆ: {request.num1} + {request.num2} = {result}",
            timestamp=int(time.time() * 1000)
        )

    def AddClientStream(self, request_iterator, context):
        """å®¢æˆ·ç«¯æµå¼RPCï¼šç´¯åŠ å¤šä¸ªæ•°å­—"""
        print(f"[{datetime.now()}] å¼€å§‹æ¥æ”¶å®¢æˆ·ç«¯æµå¼æ•°å­—...")

        total = 0
        count = 0
        numbers = []

        for number in request_iterator:
            count += 1
            total += number.value
            numbers.append(number.value)

            print(f"  æ”¶åˆ°ç¬¬{count}ä¸ªæ•°å­—: {number.value}")

        message = f"ç´¯åŠ äº† {count} ä¸ªæ•°å­—: {' + '.join(map(str, numbers))} = {total}"
        print(f"[{datetime.now()}] {message}")

        return calculator_pb2.AddResponse(
            result=total,
            message=message,
            timestamp=int(time.time() * 1000),
            request_id=f"client_stream_{int(time.time())}"
        )

    def AddBidirectional(self, request_iterator, context):
        """åŒå‘æµå¼RPCï¼šå®æ—¶è®¡ç®—"""
        print(f"[{datetime.now()}] å¼€å§‹åŒå‘æµå¼è®¡ç®—...")

        request_count = 0

        for request in request_iterator:
            request_count += 1

            # è®¡ç®—æœ¬æ¬¡è¯·æ±‚
            result = request.num1 + request.num2

            # å‘é€è®¡ç®—æ­¥éª¤
            yield calculator_pb2.CalculationStep(
                step=f"è¯·æ±‚{request_count}",
                current_result=result,
                message=f"å®æ—¶è®¡ç®—: {request.num1} + {request.num2} = {result}",
                timestamp=int(time.time() * 1000)
            )

            print(f"  å¤„ç†ç¬¬{request_count}ä¸ªè¯·æ±‚: {request.num1} + {request.num2} = {result}")

        print(f"[{datetime.now()}] åŒå‘æµå¤„ç†å®Œæˆï¼Œå…±å¤„ç† {request_count} ä¸ªè¯·æ±‚")


def serve():
    """å¯åŠ¨gRPCæœåŠ¡å™¨"""
    # åˆ›å»ºæœåŠ¡å™¨
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))

    # æ·»åŠ æœåŠ¡
    calculator_pb2_grpc.add_CalculatorServicer_to_server(
        CalculatorServicer(), server
    )

    # ç›‘å¬ç«¯å£
    port = 50051
    server.add_insecure_port(f'[::]:{port}')

    # å¯åŠ¨æœåŠ¡å™¨
    server.start()
    print(f"âœ… è®¡ç®—å™¨gRPCæœåŠ¡å·²å¯åŠ¨ï¼Œç›‘å¬ç«¯å£: {port}")
    print("ğŸ“¡ æ”¯æŒçš„RPCæ–¹æ³•:")
    print("  - Add: ç®€å•åŠ æ³•")
    print("  - AddStream: æµå¼åŠ æ³•ï¼ˆæœåŠ¡ç«¯æµï¼‰")
    print("  - AddClientStream: å®¢æˆ·ç«¯æµå¼ç´¯åŠ ")
    print("  - AddBidirectional: åŒå‘æµå¼è®¡ç®—")
    print("æŒ‰ Ctrl+C åœæ­¢æœåŠ¡å™¨...")

    try:
        # ä¿æŒè¿è¡Œ
        while True:
            time.sleep(86400)  # ä¸€å¤©
    except KeyboardInterrupt:
        print("\nğŸ›‘ æ­£åœ¨åœæ­¢æœåŠ¡å™¨...")
        server.stop(0)
        print("ğŸ‘‹ æœåŠ¡å™¨å·²åœæ­¢")


if __name__ == '__main__':
    serve()